### PR TITLE
Add better user address checks

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

With the existing logic for checking whether an address is a user address, stuff like topic indices (e.g. `0x0000000000000000000000000000000000000002`) was being flagged as a user address.

We now only consider something to be a user address if:

* It exists in the state DB
* It has a non-zero balance
* It has a non-zero nonce

(On top of the existing checks - no associated code and at least twelve leading zero bytes).

### What changes were made as part of this PR:

- Provide a high level list of the changes made
- List key areas for the reviewer 

### :rotating_light: Definition of done :rotating_light:
- [ ] Unit tests added to cover new or changed functionality 
- [ ] Docs pages updated to cover new or changed functionality
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated

### CODE REVIEW

File: go/enclave/events/subscription_manager.go [hunk 1]
```diff
@@ -169,26 +169,30 @@ func (s *SubscriptionManager) encryptLogs(logsByID map[gethrpc.ID][]*types.Log)
...
 		// (potentially) be user addresses.
-		topicHex := topic.Hex()
 		if topicHex[2:len(zeroBytesHex)+2] != zeroBytesHex {
 			continue
 		}
 
-		// If there is code associated with an address, it is not a user address.
-		addr := gethcommon.HexToAddress(topicHex)
-		if db.GetCode(addr) != nil {
+		// We also assume that a potential address is *not* a user address if any of the following is true:
```
Comment (tudor-malene) [inline go/enclave/events/subscription_manager.go:186]: it's a bit hard to follow being negative. 
I think it would e clearer if you appended in the if branch, and made the entire thing positive

**Review by tudor-malene — COMMENTED** [review]
looks good. one readability comment

File: go/enclave/events/subscription_manager.go (hunk 1 above)
Comment (joeldudley) [reply go/enclave/events/subscription_manager.go]: @tudor-malene Fixed. Can you take another look?

File: go/enclave/events/subscription_manager.go [hunk 2]
```diff
@@ -169,26 +169,27 @@ func (s *SubscriptionManager) encryptLogs(logsByID map[gethrpc.ID][]*types.Log)
...
 
 	for _, topic := range log.Topics {
-		// Since addresses are 20 bytes long, while hashes are 32, only topics with 12 leading zero bytes can
-		// (potentially) be user addresses.
-		topicHex := topic.Hex()
-		if topicHex[2:len(zeroBytesHex)+2] != zeroBytesHex {
-			continue
+		potentialAddr := gethcommon.HexToAddress(topic.Hex())
+
+		// An address is considered a user address if all the following are true:
```
Comment (tudor-malene) [inline go/enclave/events/subscription_manager.go:179]: umm, still a bit confusing. sorry.

1. start with potential address
2. check format (leading 0). Fail -> exit
3. check whether it exists in the db
 db.exist || db.balance!=0 || db.nonce !=0. 
4. If yes . Then check if the db.code is nil 
5. if yes, then append

I think it's better to express this way as it highlights the related checks

File: go/enclave/events/subscription_manager.go [hunk 3]
```diff
@@ -169,26 +169,26 @@ func (s *SubscriptionManager) encryptLogs(logsByID map[gethrpc.ID][]*types.Log)
...
-		if db.GetCode(addr) != nil {
+		// A user address must have (at least) 12 leading zero bytes, since addresses are 20 bytes long, while hashes
+		// are 32.
+		if topic.Hex()[2:len(zeroBytesHex)+2] != zeroBytesHex {
 			continue
 		}
 
-		userAddrs = append(userAddrs, addr.Hex())
+		// A user address must exist in the state DB, and have a non-zero balance and nonce.
+		if db.Exist(potentialAddr) && db.GetBalance(potentialAddr).Cmp(big.NewInt(0)) != 0 && db.GetNonce(potentialAddr) != 0 {
```
Comment (tudor-malene) [inline go/enclave/events/subscription_manager.go:186]: I think these shuold be || instead of &&

Comment (joeldudley) [reply go/enclave/events/subscription_manager.go:186]: I thought for a user address, all three had to be met? It has to exist with a real balance and nonce.

Comment (tudor-malene) [reply go/enclave/events/subscription_manager.go:186]: it is possible to have no native balance remaining, but to have submitted transactions. 
or to have balance sent by a different account, in preparation to submit transactions

Comment (joeldudley) [reply go/enclave/events/subscription_manager.go:186]: I know. But I thought we were insisting on all three? If you remember, the 0x0000002 account had a balance, but no nonce.

Comment (tudor-malene) [reply go/enclave/events/subscription_manager.go:186]: which means it's a genuine account. someone has sent money to that address

Comment (tudor-malene) [reply go/enclave/events/subscription_manager.go:186]: we are using 2 somewhere for sure. I'll find it

**Review by tudor-malene — APPROVED** [review]
lgtm
